Add support for building with msys2 mingw

schmittl opened the pull request:
For context see #133 

The changes in this PR differ slightly from the ones described in the original issue.

- Link MATH_LIBRARY only if found
- Add support for msys2 MINGW64 based builds on Windows
- Add install & build instructions to INSTALL.md
- Add compiler flags for mingw builds to CMakeLists.txt
- Check for `__MINGW32__` in markup.c so that no error is printed
- Change FindAntTweakBar.cmake to find appropriate library for 32 or 64-bit
- Change harfbuzz examples to not use symlinks
- Check if GLFW3_LIBRARY is defined before linking, alternatively use glfw for linking (I ran into the same problem as described in  #124 and tried to come up with a reasonable fix)

Note: Symlinks are not deleted yet, but harfbuzz examples are changed to not use them anymore.

I was only able to test on Windows so far, in theory it should not break other platforms.
Happy to discuss.

adrianbroher commented:
The symlinks were created intentional and with Windows related deficiencies in mind.  Please remove related changes.

> I needed to change the makefont target in CMakeLists.txt to only add MATH_LIBRARY if the library is found.

Why?

Also how is the wall of text in `INSTALL.md` related to this project?  People who choose to use mingw for whatever reason are familiar with the usage.

schmittl commented:
Hi,
thanks for the feedback.

> The symlinks were created intentional and with Windows related deficiencies in mind. Please remove related changes.

Fair enough, will update. 

> Why?

This is the error message I get without my change:

```
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
MATH_LIBRARY
    linked by target "makefont" in directory C:/git/freetype-gl
-- Configuring incomplete, errors occurred!
```

Linking the MATH_LIBRARY only if it is found is done the same way in the other CMakeLists.txt files.
But perhaps there is a better way?

> Also how is the wall of text in INSTALL.md related to this project? People who choose to use mingw for whatever reason are familiar with the usage, documentation source and download source.

True, but consider the other way around. Not everyone coming to this project knows how to get everything running on Windows. As there are already install/build instructions for Ubuntu, I figured instructions for building the demos on Windows would be a valuable contribution.  

I admit that the instructions in INSTALL.md are verbose and not only focused on freetype-gl. I will try to make them more concise.

schmittl commented:
Removed symlinks related changes.
Updated INSTALL.md to contain relevant information and instructions.

adrianbroher commented:
> Linking the MATH_LIBRARY only if it is found is done the same way in the other CMakeLists.txt files.
> But perhaps there is a better way?

I have updated the library/symbol/define detection in the master branch, can you please check if it does a better job on your build platform?

What does `__USE_MINGW_ANSI_STDIO` do?

Why do you exclude the search path defines in the main CMakeLists.txt?

schmittl commented:
> I have updated the library/symbol/define detection in the master branch, can you please check if it does a better job on your build platform?

I rebased, tested it locally and can confirm that it works for me. I will update the PR soon.

> What does __USE_MINGW_ANSI_STDIO do?

`__USE_MINGW_ANSI_STDIO` is used to provide a ANSI C99 compatible implementation of fprintf() and similar functions on MinGW. Without this define more modern format specifiers like `%zu` break, which is used in makefont.c.

> Why do you exclude the search path defines in the main CMakeLists.txt?

Without this change CMake will use the FreeType library from the Windows folder:

``` 
-- Found Freetype: C:/git/freetype-gl/windows/freetype/freetype.lib (found version "2.4.7")
```

And will then produce the following linker error:

```
Linking C executable makefont.exe
libfreetype-gl.a(texture-font.c.obj):texture-font.c:(.text+0xd4): undefined reference to `FT_Init_FreeType'

...many more undefined references.
```

With the change CMake will find and use the FreeType library from the MinGW folder and the linking works:

```
-- Found Freetype: C:/msys64/mingw64/lib/libfreetype.dll.a (found version "2.7.0")
```

I am not sure why the linker does not like the freetype.lib, but I think it makes sense to also use the  MinGW version of freetype when compiling with MinGW toolchain.

schmittl commented:
Rebased on latest master commit.
Removed the glfw linking related changes since they are now integrated in the master branch.

adrianbroher commented:
Sorry, but I need to ask you again for a rebase.  Master shouldn't need the `__USE_MINGW_ANSI_STDIO` anymore or [makefont.c:44](https://github.com/rougier/freetype-gl/blob/master/makefont.c#L44) needs a different condition.

schmittl commented:
Yup, can confirm that it works for me locally.
Rebased onto latest master commit.

adrianbroher commented:
Thanks for your contribution.
